File: src/cpp/multiprocessing/shared_array.cpp
/**
 * ⚠️  ANTIPATTERN — Shared Array Without Synchronisation
 *
 * Demonstrates anonymous shared memory (mmap MAP_ANONYMOUS) and the
 * data races that occur when multiple processes modify it concurrently
 * without any locking.
 *
 * MAP_ANONYMOUS concepts (the good part):
 *   - mmap with MAP_SHARED | MAP_ANONYMOUS creates a shared region
 *     visible to all processes that fork from the creator.
 *   - Simpler than shm_open when only parent/child access is needed.
 *   - No filesystem name; region is automatically cleaned up on exit.
 *
 * The race condition (the bad part):
 *   - ++arr[i] is not atomic: it compiles to load → increment → store.
 *   - Two processes can each load the same value, both increment, and
 *     one of the stores is lost — final value is unpredictable.
 *   - The final array values may be wrong and vary between runs.
 *
 * Fix: use a pthread_mutex_t or sem_t in shared memory to protect each
 * array element, or use a PTHREAD_PROCESS_SHARED rwlock.
 * See shared_memory.cpp for the synchronized version with a semaphore.
 *
 * Compile:
 *   g++ -std=c++17 -o shared_array shared_array.cpp
 */

#include <iostream>
#include <sys/mman.h>
#include <sys/wait.h>
#include <unistd.h>

constexpr size_t ARRAY_SIZE = 5;

void increment_array(int* arr, int iterations) {
    for (int n = 0; n < iterations; ++n) {
        for (size_t i = 0; i < ARRAY_SIZE; ++i) {
            ++arr[i];
        }
        usleep(100'000);
    }
}

void decrement_array(int* arr, int iterations) {
    for (int n = 0; n < iterations; ++n) {
        for (size_t i = 0; i < ARRAY_SIZE; ++i) {
            --arr[i];
        }
        usleep(100'000);
    }
}

void print_array(const char* label, int* arr) {
    std::cout << label << ": [";
    for (size_t i = 0; i < ARRAY_SIZE; ++i) {
        std::cout << arr[i];
        if (i < ARRAY_SIZE - 1) std::cout << ", ";
    }
    std::cout << "]\n";
}

int main() {
    // Create anonymous shared memory for the array
    int* shared_array = static_cast<int*>(
        mmap(nullptr, sizeof(int) * ARRAY_SIZE,
             PROT_READ | PROT_WRITE,
             MAP_SHARED | MAP_ANONYMOUS, -1, 0));

    // Initialize array
    for (size_t i = 0; i < ARRAY_SIZE; ++i) {
        shared_array[i] = 0;
    }

    print_array("Initial", shared_array);

    pid_t pid1 = fork();
    if (pid1 == 0) {
        increment_array(shared_array, 3);
        print_array("After increment", shared_array);
        _exit(0);
    }

    pid_t pid2 = fork();
    if (pid2 == 0) {
        decrement_array(shared_array, 3);
        print_array("After decrement", shared_array);
        _exit(0);
    }

    // Wait for children
    waitpid(pid1, nullptr, 0);
    waitpid(pid2, nullptr, 0);

    print_array("Final", shared_array);

    munmap(shared_array, sizeof(int) * ARRAY_SIZE);
    return 0;
}

File: src/python/multiprocessing/shared_array.py
"""
Shared Array Example

This script demonstrates sharing an array between processes using
multiprocessing.Array. This provides fast shared memory access for
sequences of primitive values.

Key Concepts:
- Array creates a ctypes array in shared memory
- Type codes: 'i' (int), 'd' (double), 'f' (float), etc.
- Access elements using index notation or iteration
- Use a Lock for process-safe modification
- Protect a whole logical update, not each element separately

Note:
This example runs two processes that modify the same array for a fixed
number of iterations. The lock protects each full-array update so another
process cannot observe a half-updated array.

For numpy arrays, consider multiprocessing.shared_memory (Python 3.8+)
or numpy.memmap for more efficient shared array operations.
"""

import multiprocessing
import time


def increment_array(shared_array, lock, num_iterations=5):
    """Increment all elements in the shared array."""
    for _ in range(num_iterations):
        with lock:
            for i in range(len(shared_array)):
                shared_array[i] += 1
            print(f"Array incremented: {list(shared_array)}")
        time.sleep(1)


def decrement_array(shared_array, lock, num_iterations=5):
    """Decrement all elements in the shared array."""
    for _ in range(num_iterations):
        with lock:
            for i in range(len(shared_array)):
                shared_array[i] -= 1
            print(f"Array decremented: {list(shared_array)}")
        time.sleep(1)


def main():
    # Create shared array of integers initialized to [0, 0, 0]
    shared_array = multiprocessing.Array("i", [0, 0, 0], lock=False)
    lock = multiprocessing.Lock()

    process1 = multiprocessing.Process(
        target=increment_array, args=(shared_array, lock, 5)
    )
    process2 = multiprocessing.Process(
        target=decrement_array, args=(shared_array, lock, 5)
    )

    process1.start()
    process2.start()

    process1.join()
    process2.join()

    print(f"Final array state: {list(shared_array)}")


if __name__ == "__main__":
    main()
